### ISSUE TITLE
Copy and paste takes 12 seconds to complete

### ISSUE THREAD

**Issue by KoalaBear84** [ffMathy/Shapeshifter#349]
[CopyThisContent.txt](https://github.com/ffMathy/Shapeshifter/files/1963841/CopyThisContent.txt)

If you copy the contents of this, it thinks that there are URLs in it, which it all tries to resolve. Also the Regex can't complete within time giving errors.

Example errors:
`System.Net.Sockets.SocketException (0x80004005): No such host is known`

`System.Text.RegularExpressions.RegexMatchTimeoutException: The RegEx engine has timed out while trying to match a pattern to an input string. This can occur for many reasons, including very large inputs or excessive backtracking caused by nested quantifiers, back-references and other factors.`

There are probably two things that needs to be done:
1. Make all copy parsers / things async, don't lock / capture the clipboard, so the user can still paste it
2. Better / faster regex (with async a little less important)
3. Distinct list of URLs, so it might only be one url a 100 times or so (I didn't check how it is implemented right now)

**Comment by KoalaBear84** [comment]
I see it checks `IsValidDomainAsync` a lot of times for "Shapeshifter.exe" 😇

**Comment by ffMathy** [comment]
We should use https://github.com/0xcb/Re2.Net instead of C# Regular Expressions. RE2 is Google's regular expression library which is guaranteed to not go to an infinite loop and is secure to use.

The parsers should already be async, so that's a mistake if they are not.

**Comment by KoalaBear84** [comment]
Yes, I see the whole parsing in itself is async, and also with yield, very good. I think it just captures the clipboard also while doing all this in async. So you cannot use the clipboard until Shapeshifter is done.

**Comment by ffMathy** [comment]
The thing is that we can never know if something is a website or not. So what Shapeshifter will try to do is to make a DNS resolve of all words in the clipboard to detect possible URLs to open. That should be very fast.

The errors you are seeing, are you sure they aren't just "first chance" exceptions? Those are exceptions that are actually caught, but displayed anyway while debugging.

**Comment by KoalaBear84** [comment]
You are right about the first chance exceptions, and also luckily, else we have a lot of issues on Github 🚀

**Comment by ffMathy** [comment]
They have no NuGet package apparently. I have asked them to create one.

https://github.com/0xcb/Re2.Net/issues/4

**Comment by ffMathy** [comment]
@KoalaBear84 I have trimmed the logs heavily in the latest pre-release that is up in half an hour or so (or you can pull latest `develop` now and see). I also disabled first-chance exceptions from popping up.

See if it makes debugging easier, and thank you so much for helping me debug!

**Comment by ffMathy** [comment]
This should be switched to the time of copying and then persisted, instead of being generated dynamically upon switching to the item. That would solve a lot.

**Comment by ffMathy** [comment]
I now greatly improved the regex performance by switching to RE2, but still leaving this here, as generating actions should never delay clipboard performance.

**Comment by ffMathy** [comment]
While that may be faster, it is only faster in the scenario where you don't switch. I would much rather force the data through the same funnel always for consistency.

Here's what I'll do:
1. When copying something, it'll start analyzing the data in the background to determine what kind of actions are available for that data. It will do this in an asynchronously queued manner.
2. Once the analysis for an action is done, that action gets added to the list of available actions for that item. 
3. When the user shows the UI, we only display the actions that have finished analyzing for each item. So if one action is slow to analyze, the user just won't see that action. 

This way, the user can always copy and switch right away, and because the data is not analyzed upon showing the UI but instead in the background when a new item has been copied, it should also feel faster. 

Does that make sense?

**Comment by KoalaBear84** [comment]
Yes. I think point 2 is already working, at least a little. When I tried it back then, the 'Open URLs' action was not yet shown, but after a couple of seconds it WAS, dynamically, while the overlay was open.

**Comment by ffMathy** [comment]
Can you see if the latest prerelease improves the performance? It really should. Issue should still be present though.

**Comment by ffMathy** [comment]
Hmmm it is actually equally bad if not worse now. I opened #541 in addition to this to improve upon it.

**Comment by ffMathy** [comment]
All link related actions have now been entirely disabled until we stabilize all of this. There is no reason for it degrading the experience given how much they are used.

Prerelease up in a few minutes.

**Comment by ffMathy** [comment]
@KoalaBear84 how is the latest prerelease performing? It should be significantly better, but there is still work to be done.

**Comment by ffMathy** [comment]
This has now been fixed. Deferred action resolving has been implemented, and it's insanely fast and effective.

**Comment by KoalaBear84** [comment]
Great. Will check it out.

**Comment by ffMathy** [comment]
Be sure to download it manually though because latest prereleases still have issues auto updating.

**Comment by KoalaBear84** [comment]
Hmm, I did. shapeshifter-v6.2018.606.2854. It instantly crashes.

Also cleaned `C:\Program Files\Shapeshifter` and also tried to put Shapeshifter.exe directly in the right folder. But unfortunately nothing helps, cannot start it. Crash dump is generated.

**Comment by ffMathy** [comment]
Running it from the installation folder is not a good idea. Try cleaning the installation folder and removing it entirely, then run the downloaded file.

**Comment by KoalaBear84** [comment]
Yes, I did that. Hmm, this is strange. But now I also deleted `C:\Program Files\Shapeshifter` and it works.

**Comment by KoalaBear84** [comment]
Hmm. I just copied 193 kB of a Shapeshifter log file. And eventually it showed the "Shapeshifter is not responding".

`2018-06-08 12:14:57.115 [#15868] [Shapeshifter.WindowsDesktop.Infrastructure.Handles.Factories.PerformanceHandleFactory] [VRB]
Finished executing HasLinkWithConditionsAsync in 6589.7254 milliseconds.

2018-06-08 12:15:03.581 [#15868] [Shapeshifter.WindowsDesktop.Infrastructure.Handles.Factories.PerformanceHandleFactory] [VRB]
Finished executing HasLinkWithConditionsAsync in 6466.6556 milliseconds.

2018-06-08 12:15:10.127 [#15868] [Shapeshifter.WindowsDesktop.Infrastructure.Handles.Factories.PerformanceHandleFactory] [VRB]
Finished executing HasLinkWithConditionsAsync in 6545.4271 milliseconds.

2018-06-08 12:16:06.474 [#15868] [Shapeshifter.WindowsDesktop.Infrastructure.Handles.Factories.PerformanceHandleFactory] [VRB]
Finished executing HasLinkWithConditionsAsync in 56346.9137 milliseconds.`
